File: backend/app/schemas/tournament.py
"""Tournament DTOs."""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field

from app.domain.rules import MatchRules
from app.schemas.roster import PlayerDTO


class TournamentCreate(BaseModel):
    name: str = Field(..., min_length=1)
    format: str = Field(default="round_robin", pattern="^(round_robin|knockout|groups)$")
    team_ids: list[str] = Field(..., min_length=2)
    format_id: str = Field(default="t20", description="preset for the matches' rules")
    rules: Optional[MatchRules] = Field(default=None, description="custom rules override")
    # 'groups' format: pools + a seeded knockout playoff
    num_groups: int = Field(default=2, ge=2, le=16)
    advance_per_group: int = Field(default=2, ge=1, le=8)
    # configurable points (apply to league + groups standings)
    win_points: int = Field(default=2, ge=0, le=20)
    tie_points: int = Field(default=1, ge=0, le=20)
    nr_points: int = Field(default=1, ge=0, le=20)
    # rain rules for every match in the tournament (DLS revised targets)
    dls_enabled: bool = False


class TournamentSettings(BaseModel):
    """Admin-toggleable tournament settings (currently just the DLS switch)."""

    dls_enabled: bool


class TeamRef(BaseModel):
    id: str
    name: str


class FixtureDTO(BaseModel):
    id: str
    round: int
    position: int
    team_a: Optional[TeamRef]
    team_b: Optional[TeamRef]
    match_id: Optional[str]
    status: str  # scheduled | live | completed
    result: Optional[str] = None
    group: Optional[str] = None  # pool label, or None for a playoff/bracket fixture


class StandingRow(BaseModel):
    team_id: str
    name: str
    played: int = 0
    won: int = 0
    lost: int = 0
    tied: int = 0
    no_result: int = 0
    points: int = 0
    nrr: float = 0.0


class TournamentDTO(BaseModel):
    id: str
    name: str
    format: str
    status: str
    teams: list[TeamRef]


class GroupStanding(BaseModel):
    """One pool's points table in a 'groups' tournament."""

    group: str  # "A", "B", …
    standings: list[StandingRow] = []


class TournamentDetailDTO(TournamentDTO):
    fixtures: list[FixtureDTO]
    standings: list[StandingRow]  # flat league table (round_robin); empty for groups
    groups: list[GroupStanding] = []  # per-pool tables (groups format)
    champion: Optional[TeamRef] = None
    config: dict = Field(default_factory=dict)  # points + group settings echo
    is_manager: bool = Field(
        default=False,
        description="May the caller run this competition (its organizer, or an admin)?",
    )
    """Viewer-aware, and the answer to a question the client previously had to
    ask by attacking the API: nothing on the detail said who manages a
    competition, so the app probed a management endpoint and read the 403 as
    "not yours" — one refused request on every visit by everybody else.

    Always ``False`` unless a route sets it from the authenticated caller via
    ``ScopeService``. It is never read from the request, and it is a hint for
    the UI: the server still refuses the write, whatever the client was told.
    """


class StartFixtureRequest(BaseModel):
    squad_a_ids: list[str] = Field(..., min_length=2)
    squad_b_ids: list[str] = Field(..., min_length=2)
    bat_first: str = Field(default="a", pattern="^(a|b)$")


class TeamSquadDTO(BaseModel):
    """A team's registered squad for one tournament."""

    team_id: str
    team_name: str
    players: list[PlayerDTO] = Field(default_factory=list)


class SquadRegisterRequest(BaseModel):
    player_id: str


class PlayerTournamentDTO(BaseModel):
    """One competition a player is *registered* in, from the other side.

    Distinct from the tournament buckets in a player's history: those are
    derived from matches already played, so a player entered into a squad but
    not yet fielded appears in none of them. This is the registration itself,
    which exists from the moment the organizer picks them — with ``has_played``
    saying which of the two facts is true.
    """

    tournament_id: str
    tournament_name: str
    format: str
    status: str
    team_id: str
    team_name: str
    has_played: bool = False
    matches_played: int = 0

File: mobile/lib/core/models/tournament.dart
/// Tournaments — leagues, knockouts and group-stage competitions.
library;

import 'json.dart';
import 'roster.dart';

class TeamRef {
  final String id;
  final String name;

  const TeamRef({required this.id, required this.name});

  factory TeamRef.fromJson(Map<String, dynamic> j) =>
      TeamRef(id: asStr(j['id']), name: asStr(j['name']));
}

/// Competition shapes the backend accepts.
class TournamentFormats {
  const TournamentFormats._();
  static const roundRobin = 'round_robin';
  static const knockout = 'knockout';
  static const groups = 'groups';

  static const all = <String>[roundRobin, groups, knockout];

  static String label(String v) => switch (v) {
        roundRobin => 'League',
        knockout => 'Knockout',
        groups => 'Groups + playoffs',
        _ => v,
      };

  static String blurb(String v) => switch (v) {
        roundRobin => 'Everyone plays everyone. Points table with net run rate.',
        knockout => 'Single elimination. Lose once and you are out.',
        groups => 'Pools first, then a seeded knockout playoff.',
        _ => '',
      };
}

class Fixture {
  final String id;
  final int round;
  final int position;
  final TeamRef? teamA;
  final TeamRef? teamB;
  final String? matchId;

  /// scheduled | live | completed
  final String status;
  final String? result;

  /// Pool label ("A", "B", …); null for a playoff or bracket fixture.
  final String? group;

  const Fixture({
    required this.id,
    this.round = 1,
    this.position = 1,
    this.teamA,
    this.teamB,
    this.matchId,
    this.status = 'scheduled',
    this.result,
    this.group,
  });

  factory Fixture.fromJson(Map<String, dynamic> j) => Fixture(
        id: asStr(j['id']),
        round: asInt(j['round'], 1),
        position: asInt(j['position'], 1),
        teamA: j['team_a'] is Map ? TeamRef.fromJson(asMap(j['team_a'])) : null,
        teamB: j['team_b'] is Map ? TeamRef.fromJson(asMap(j['team_b'])) : null,
        matchId: asStrOrNull(j['match_id']),
        status: asStr(j['status'], 'scheduled'),
        result: asStrOrNull(j['result']),
        group: asStrOrNull(j['group']),
      );

  bool get isCompleted => status == 'completed';
  bool get isLive => status == 'live';
  bool get isScheduled => status == 'scheduled';

  /// A team drawing a bye has no opponent.
  bool get isBye => result == 'bye' || teamB == null;

  String get title {
    final a = teamA?.name ?? 'TBD';
    final b = teamB?.name ?? (isBye ? 'Bye' : 'TBD');
    return '$a v $b';
  }
}

class StandingRow {
  final String teamId;
  final String name;
  final int played;
  final int won;
  final int lost;
  final int tied;
  final int noResult;
  final int points;
  final double nrr;

  const StandingRow({
    required this.teamId,
    this.name = '',
    this.played = 0,
    this.won = 0,
    this.lost = 0,
    this.tied = 0,
    this.noResult = 0,
    this.points = 0,
    this.nrr = 0,
  });

  factory StandingRow.fromJson(Map<String, dynamic> j) => StandingRow(
        teamId: asStr(j['team_id']),
        name: asStr(j['name']),
        played: asInt(j['played']),
        won: asInt(j['won']),
        lost: asInt(j['lost']),
        tied: asInt(j['tied']),
        noResult: asInt(j['no_result']),
        points: asInt(j['points']),
        nrr: asDouble(j['nrr']),
      );

  /// Net run rate with an explicit sign, the way points tables print it.
  String get nrrText =>
      '${nrr >= 0 ? '+' : ''}${nrr.toStringAsFixed(3)}';
}

class GroupStanding {
  final String group;
  final List<StandingRow> standings;

  const GroupStanding({required this.group, this.standings = const []});

  factory GroupStanding.fromJson(Map<String, dynamic> j) => GroupStanding(
        group: asStr(j['group']),
        standings: asMapList(j['standings']).map(StandingRow.fromJson).toList(),
      );
}

class TournamentSummary {
  final String id;
  final String name;
  final String format;
  final String status;
  final List<TeamRef> teams;

  const TournamentSummary({
    required this.id,
    this.name = '',
    this.format = TournamentFormats.roundRobin,
    this.status = 'active',
    this.teams = const [],
  });

  factory TournamentSummary.fromJson(Map<String, dynamic> j) =>
      TournamentSummary(
        id: asStr(j['id']),
        name: asStr(j['name']),
        format: asStr(j['format'], TournamentFormats.roundRobin),
        status: asStr(j['status'], 'active'),
        teams: asMapList(j['teams']).map(TeamRef.fromJson).toList(),
      );

  String get subtitle =>
      '${teams.length} teams · ${TournamentFormats.label(format)}';
}

class Tournament {
  final String id;
  final String name;
  final String format;
  final String status;
  final List<TeamRef> teams;
  final List<Fixture> fixtures;

  /// Flat league table (round_robin); empty for groups and knockout.
  final List<StandingRow> standings;

  /// Per-pool tables (groups format).
  final List<GroupStanding> groups;
  final TeamRef? champion;

  /// Echo of the points and group settings.
  final Map<String, dynamic> config;

  const Tournament({
    required this.id,
    this.name = '',
    this.format = TournamentFormats.roundRobin,
    this.status = 'active',
    this.teams = const [],
    this.fixtures = const [],
    this.standings = const [],
    this.groups = const [],
    this.champion,
    this.config = const {},
  });

  factory Tournament.fromJson(Map<String, dynamic> j) => Tournament(
        id: asStr(j['id']),
        name: asStr(j['name']),
        format: asStr(j['format'], TournamentFormats.roundRobin),
        status: asStr(j['status'], 'active'),
        teams: asMapList(j['teams']).map(TeamRef.fromJson).toList(),
        fixtures: asMapList(j['fixtures']).map(Fixture.fromJson).toList(),
        standings: asMapList(j['standings']).map(StandingRow.fromJson).toList(),
        groups: asMapList(j['groups']).map(GroupStanding.fromJson).toList(),
        champion: j['champion'] is Map
            ? TeamRef.fromJson(asMap(j['champion']))
            : null,
        config: asMap(j['config']),
      );

  bool get isKnockout => format == TournamentFormats.knockout;
  bool get isGroups => format == TournamentFormats.groups;
  bool get isLeague => format == TournamentFormats.roundRobin;

  bool get dlsEnabled => asBool(config['dls_enabled']);

  /// Bracket fixtures — knockout rounds, or the playoffs of a groups event.
  List<Fixture> get bracketFixtures =>
      fixtures.where((f) => f.group == null).toList();

  List<Fixture> fixturesInGroup(String group) =>
      fixtures.where((f) => f.group == group).toList();

  /// Fixtures bucketed by round, in round order.
  Map<int, List<Fixture>> roundsOf(List<Fixture> source) {
    final out = <int, List<Fixture>>{};
    for (final f in source) {
      out.putIfAbsent(f.round, () => []).add(f);
    }
    return Map.fromEntries(
      out.entries.toList()..sort((a, b) => a.key.compareTo(b.key)),
    );
  }

  /// Knockout rounds get named from the end: Final, Semi-finals, and so on.
  static String roundLabel(int round, int totalRounds, {bool knockout = true}) {
    if (!knockout) return 'Round $round';
    final fromEnd = totalRounds - round;
    return switch (fromEnd) {
      0 => 'Final',
      1 => 'Semi-finals',
      2 => 'Quarter-finals',
      _ => 'Round $round',
    };
  }
}

/// A team's registered squad for one tournament.
class TeamSquad {
  final String teamId;
  final String teamName;
  final List<Player> players;

  const TeamSquad({
    required this.teamId,
    this.teamName = '',
    this.players = const [],
  });

  factory TeamSquad.fromJson(Map<String, dynamic> j) => TeamSquad(
        teamId: asStr(j['team_id']),
        teamName: asStr(j['team_name']),
        players: asMapList(j['players']).map(Player.fromJson).toList(),
      );
}
